rep-2012.rst

```diff
@@ -0,0 +1,433 @@
+  A ROS service client.
+  Sends requests to services and receives responses.
+
+Motivation
+==========
+
+The primary motivation for this proposal is to make it easier for users to externally validate that services are operating as expected.
+Drawing an analogy to ROS topics, there exists tools and libraries for "echoing" and recording messages sent over a topic, and we would like to see the same kind of capabilities for services.
+Specifically, we want the capability to introspect requests and responses that are sent to and from services and clients.
+Being able to remotely monitor services allows users to more effectively troubleshoot issues in a ROS system.
+For example, we could verify requests are being received by a service by employing a command-line tool during runtime.
+Or, we could post-process recorded requests and responses to validate their content.
+
+There are additional features that could leverage this proposal, such as:
+
+- Playback of recorded services (for example, from a rosbag [1]_)
+- Introspection of ROS actions, which are built on services
+- Validation of a live ROS system by referencing a recording from a previous session
+
+Though this proposal focuses on the core feature of introspecting requests and responses, we want to keep the design flexible so additional features like those listed above can be implemented in the future.
+
+
+Specification
+=============
+
+Publishing Service Events
+-------------------------
+
+Whenever a request or response is sent or received, a *service event* message will be published to a topic.
+Services are responsible for publishing a message when they receive a request and when they send a response.
+Likewise, clients are responsible for publishing a message when they send a request and when they receive a response.
+Therefore, we have a total of four possible events:
+
+:Request Sent:
+  Emitted from a client after sending a request to a service.
+:Request Received:
+  Emitted from a service after receiving a request from a client.
+:Response Sent:
+  Emitted from a service after sending a response to a client.
+:Response Received:
+  Emitted from a client after receiving a response from a service.
+
+Event messages shall be published to the hidden topic ``/SERVICE_NAME/_service_event``, where ``SERVICE_NAME`` is the fully-qualified name of the service.
+Both services and clients will publish events to the same topic.
+Note that this implies that services must have unique names.
+
+By publishing service event messages to predetermined topics, tools and libraries are able to subscribe to these topics to inspect the flow of data between services.
+
+Service Event Definition
+------------------------
+
+For each service definition, ``my/srv/Foo.srv``, we define a new ROS message type, ``my/srv/Foo_Event.msg``, with the ROS IDL specification [2]_:
+
+  .. code-block::
+
+   # Event info
+   # Contains event type, timestamp, and request ID
+   rcl_interfaces/msg/ServiceEventInfo info
```

Comment:
A drawback of defining the metadata in a separate message type is that we now have a kind of dependency cycle between the code generation pipeline (`rosidl`) and this message type. I.e. `rcl_interfaces/msg/ServiceEventInfo` depends on the `rosidl` pipeline, and `rosidl` depends on `rcl_interfaces/msg/ServiceEventInfo` for generating code to support ROS services.

We have a similar problem with ROS Actions, which depend on `action_msgs`. At the moment, this adds a constraint on users defining custom Actions to declare a dependency on `action_msgs`, otherwise [CMake will fail with an error](https://github.com/ros2/rosidl/blob/51811962c832745f531968d351af3d1af45cd238/rosidl_cmake/cmake/rosidl_generate_interfaces.cmake#L145-L151). We could enforce a similar constraint on building custom services, producing a CMake error if users do not declare a dependency on `rcl_interfaces` (or maybe a new `service_msgs`), but this is likely to break a number of existing packages.

It might be better to instead define the service event metadata with primitive types like,

```
   uint8 REQUEST_SENT      = 0
   uint8 REQUEST_RECEIVED  = 1
   uint8 RESPONSE_SENT     = 2
   uint8 RESPONSE_RECEIVED = 3

   uint8 event_type

   int32 stamp_sec
   uint32 stamp_nanpsec

   uint8[16] client_id

   int64 sequence_number
```

Reply:
Hmm, that sucks.

It seems like both the Action issue and this issue could be solved by having a separate package for actual `rosidl` generation, and a "facade" one that we recommend people depend on. Packages like `builtin_interfaces`, `action_msgs`, and maybe a new `service_msgs` package could depend on the "internal" `rosidl` generator, and everyone else could depend on the "normal" `rosidl` generator package which itself would depend on the "internal" one and any necessary additional message packages.

`rviz` does something like this where:

- `rviz2` is a mostly empty package which depends on `rviz_default_plugins` and `rviz_common`,
- `rviz_default_plugins` depends on `rviz_common`, and
- `rviz_common` is the majority of the implementation.

Let me know if this makes sense. I can talk with you more about it off-line if you like.

Reply:
I think this issue has been resolved. We separated the existing meta-package (`rosidl_defaults`) into two packages, one of which the core packages like `action_msgs` and `service_msgs` can depend on, while the other users can continue to depend on. See https://github.com/ros2/rosidl_defaults/pull/22  and connected PRs for more details.